bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/EcovacsXmppDevice.java

```diff
@@ -0,0 +1,463 @@
+                            }
+                        }
+                    } else {
+                        Optional<String> eventNameOpt = XPathUtils.getFirstXPathMatchOpt(iq.payload, "//ctl/@td")
+                                .map(n -> n.getNodeValue());
+                        if (eventNameOpt.isPresent()) {
+                            logger.trace("{}: Received event message XML {}", getSerialNumber(), iq.payload);
+                            parser.handleMessage(eventNameOpt.get(), iq.payload);
+                        } else {
+                            logger.debug("{}: Got unexpected XML payload {}", getSerialNumber(), iq.payload);
+                        }
+                    }
+                } catch (Exception e) {
+                    listener.onEventStreamFailure(EcovacsXmppDevice.this, e);
+                }
+            } else if (iqRequest instanceof ErrorIQ) {
+                StanzaError error = ((ErrorIQ) iqRequest).getError();
+                logger.trace("{}: Got error response {}", getSerialNumber(), error);
+                listener.onEventStreamFailure(EcovacsXmppDevice.this,
+                        new XMPPException.XMPPErrorException(iqRequest, error));
+            }
+            return null;
+        }
+    }
+
+    private static class CommandResponseHolder {
+        @Nullable
+        String response;
+    }
+
+    private static class DeviceCommandIQ extends IQ {
+        static final String TAG_NAME = "query";
+        static final String NAMESPACE = "com:ctl";
+
+        private final String payload;
+        final String id;
+
+        // request
+        public DeviceCommandIQ(IotDeviceCommand<?> cmd, Jid from, Jid to) throws Exception {
+            super(TAG_NAME, NAMESPACE);
+            setType(Type.set);
+            setTo(to);
+            setFrom(from);
+
+            this.id = createRequestId();
+            this.payload = cmd.getXmlPayload(id);
+        }
+
+        // response
+        public DeviceCommandIQ(@Nullable String id, String payload) {
+            super(TAG_NAME, NAMESPACE);
+            this.id = id != null ? id : "";
+            this.payload = payload.replaceAll("\n|\r", "");
+        }
+
+        @Override
+        protected @Nullable IQChildElementXmlStringBuilder getIQChildElementBuilder(
+                @Nullable IQChildElementXmlStringBuilder xml) {
+            if (xml != null) {
+                xml.rightAngleBracket();
+                xml.append(payload);
+            }
+            return xml;
+        }
+
+        private String createRequestId() {
+            // Ecovacs' app uses numbers for request IDs, so better constrain ourselves to that as well
+            int random8DigitNumber = 10000000 + new Random().nextInt(90000000);
+            return Integer.toString(random8DigitNumber);
+        }
+    }
+
+    private static class PingIQ extends SimpleIQ {
+        public PingIQ(Jid from, Jid to) {
```

Comment:
Why not use org.jivesoftware.smackx.ping.packet.Ping?

Reply:
Because smack-extensions depends on smack-im, which I didn't want to pull in due to the issues mentioned above. I couldn't use `PingManager`s timer because I want to ping the device (a particular JID) and not the server [1].
Therefore I thought duplicating that simple class should be fine.

[1] Ecovacs implements an XMPP proxy server that itself talks to the actual device, and while the connection to their server is stable, the connection between their server and the device looks like it isn't.

Reply:
Since I now need `smack-im` for disabling the roster loading, I also switched to using `PingManager`.